lib/mongoid/persistable/settable.rb

```diff
@@ -23,17 +51,39 @@ def set(setters)
         prepare_atomic_operation do |ops|
           process_atomic_operations(setters) do |field, value|
 
-            field_and_value_hash = hasherizer(field.split('.'), value)
-            field = field_and_value_hash.keys.first.to_s
+            field_seq = field.to_s.split('.')
+            field = field_seq.shift
+            if field_seq.length > 0
+              # nested hash path
+              old_value = attributes[field]
+
+              # if the old value is not a hash, clobber it
+              unless Hash === old_value
+                old_value = {}
+              end
+
+              # descend into the hash, creating intermediate keys as needed
+              cur_value = old_value
+              while field_seq.length > 1
+                cur_key = field_seq.shift
+                # clobber on each level if type is not a hash
+                unless Hash === cur_value[cur_key]
+                  cur_value[cur_key] = {}
+                end
+                cur_value = cur_value[cur_key]
+              end
 
-            if fields[field] && fields[field].type == Hash && attributes.key?(field)
-              merger = proc { |key, v1, v2| Hash === v1 && Hash === v2 ? v1.merge(v2, &merger) : v2 }
-              value = (attributes[field] || {}).merge(field_and_value_hash[field], &merger)
-              process_attribute(field.to_s, value)
-            else
-              process_attribute(field.to_s, field_and_value_hash[field])
+              # now we are on the leaf level, perform the set
+              # and overwrite whatever was on this level before
+              cur_value[field_seq.shift] = value
+
+              # and set value to the value of the top level field
+              # because this is what we pass to $set
+              value = old_value
```

Comment:
now that the bulk of this `set` method consists of logic handling nested hashing, perhaps section should be pulled out into a separate method specifically for merging hashes?